subprocess/cmd.go

```diff
@@ -0,0 +1,110 @@
+package subprocess
+
+import (
+	"io"
+	"os"
+	"os/exec"
+)
+
+// Thin wrapper around exec.Cmd. Takes care of pipe shutdown by
+// keeping an internal reference to any created pipes. Whenever
+// Cmd.Wait() is called, all created pipes are closed.
+type SafeCmd struct {
+	Path string
+	Args []string
+	Env  []string
+	Dir  string
+
+	Stdin  io.Reader
+	Stdout io.Writer
+	Stderr io.Writer
+
+	Process *os.Process
+
+	cmd        *exec.Cmd
+	stdoutPipe io.ReadCloser
+	stderrPipe io.ReadCloser
+	stdinPipe  io.WriteCloser
+}
+
+func (cw *SafeCmd) StdoutPipe() (io.ReadCloser, error) {
+	stdout, err := cw.cmd.StdoutPipe()
+	cw.stdoutPipe = stdout
+	return stdout, err
+}
+
+func (cw *SafeCmd) StderrPipe() (io.ReadCloser, error) {
+	stderr, err := cw.cmd.StderrPipe()
+	cw.stderrPipe = stderr
+	return stderr, err
+}
+
+func (cw *SafeCmd) StdinPipe() (io.WriteCloser, error) {
+	stdin, err := cw.cmd.StdinPipe()
+	cw.stdinPipe = stdin
+	return stdin, err
+}
+
+func (cw *SafeCmd) Output() ([]byte, error) {
+	cw.assignIO()
+	outp, err := cw.cmd.Output()
+	cw.Process = cw.cmd.Process
+	return outp, err
+}
+
+func (cw *SafeCmd) CombinedOutput() ([]byte, error) {
+	cw.assignIO()
+	outp, err := cw.cmd.CombinedOutput()
+	cw.Process = cw.cmd.Process
+	return outp, err
+}
+
+func (cw *SafeCmd) Start() error {
+	cw.assignIO()
+	err := cw.cmd.Start()
+	cw.Process = cw.cmd.Process
+	return err
+}
+
+func (cw *SafeCmd) Run() error {
+	cw.assignIO()
+	err := cw.cmd.Run()
+	cw.Process = cw.cmd.Process
+	return err
+}
+
+func (cw *SafeCmd) assignIO() {
+	if cw.Stdin != nil {
+		cw.cmd.Stdin = cw.Stdin
+	}
+	if cw.Stdout != nil {
+		cw.cmd.Stdout = cw.Stdout
+	}
+	if cw.Stderr != nil {
+		cw.cmd.Stderr = cw.Stderr
+	}
+}
+
+func (cw *SafeCmd) Wait() error {
```

Comment:
I think the `*Pipe()` functions need to be overridden as well. There's no way to get ahold of the pipes created by `exec.Cmd`'s functions.